spec/weather_retriever_spec.rb

```diff
@@ -0,0 +1,149 @@
+# frozen_string_literal: true
+
+require 'open_weather_client'
+require 'city_name_converter'
+require 'sql_database'
+require 'weather_retriever'
+
+RSpec.describe WeatherRetriever do
+  let(:presenter) { double('presenter') }
+  let(:sql) { SQLite3::Database.open('file::memory:?cache=shared') }
+  let(:weather_info) { File.open('fixtures/london_weather.json').read }
+
+  before do
+    allow(Date).to receive(:today).and_return Date.new(2021, 4, 9)
+    open_weather_stub
+    presenter_stub
+  end
+
+  after { sql.execute 'DROP TABLE weather' }
+
+  describe '#run' do
+    it 'collects the city name' do
+      set_up_retriever.run
+
+      expect(presenter).to have_received(:city_name).once
+    end
+
+    it 'does not call the client if the database is up to date' do
+      retriever = set_up_retriever
+
+      add_weather_row
+      retriever.run
+
+      expect(open_weather_stub).not_to have_been_made
+    end
+
+    context 'calls the client' do
+      it 'if database is empty' do
+        set_up_retriever.run
+
+        expect(open_weather_stub).to have_been_made.once
+      end
+
+      it 'if database is out of date' do
+        yesterday = 1_617_896_069
+        retriever = set_up_retriever
+
+        add_weather_row(yesterday)
+        retriever.run
+
+        expect(open_weather_stub).to have_been_made.once
+      end
```

Comment:
Just so that you don't have a reference to an obscure random number, it's better to use clear names. You could have a `today` let variable defined like this:

```ruby
let(:today) { Time.utc(2021, 4, 9) }
```

and then replace all the references to the arcane number that represents `today` in seconds with your new variable (and update the "dt" field in the JSON fixture). To get the seconds from it you just call `to_i`:

```ruby
today.to_i
```

And finally, if you want to check that it is in the range of a day, the code should not compare with yesterday, but rather with two days ago.
I would also pass the data to `add_weather_row` as a named argument instead.
So putting it all together things are super explicit and readable:

```ruby
two_days_ago = Time.utc(today.year, today.month, today.day - 2)
add_weather_row(unix_date: two_days_ago.to_i)

# ...etc.

def add_weather_row(unix_date: today.to_i)
# ...etc.
```